### PR TITLE
[data] [streaming] Enable actor fault tolerance by default for streaming map operator

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Enable actor restarts (infinite) and also task retries (up to 5 by default).

### CODE REVIEW

File: python/ray/data/_internal/execution/operators/actor_pool_map_operator.py [hunk 1]
```diff
@@ -291,6 +291,16 @@ def _apply_default_remote_args(ray_remote_args: Dict[str, Any]) -> Dict[str, Any
                 ray_remote_args["scheduling_strategy"] = "SPREAD"
             else:
                 ray_remote_args["scheduling_strategy"] = ctx.scheduling_strategy
+        # Enable actor fault tolerance by default, with infinite actor recreations and
+        # up to three retries per task. The user can customize this in map_batches via
```
Comment (c21) [inline python/ray/data/_internal/execution/operators/actor_pool_map_operator.py:295]: should it be `five retries`?

File: python/ray/data/_internal/execution/operators/actor_pool_map_operator.py [hunk 2]
```diff
@@ -291,6 +291,16 @@ def _apply_default_remote_args(ray_remote_args: Dict[str, Any]) -> Dict[str, Any
...
+        # extra kwargs (e.g., map_batches(..., max_restarts=0) to disable).
+        if "max_restarts" not in ray_remote_args:
+            ray_remote_args["max_restarts"] = -1
```
Comment (c21) [inline python/ray/data/_internal/execution/operators/actor_pool_map_operator.py:298]: is there any reason why we want to restart actor infinitely? What if user code has bug, and causes the infinite retry and stuck?

File: python/ray/data/_internal/execution/operators/actor_pool_map_operator.py [hunk 3]
```diff
@@ -291,6 +291,16 @@ def _apply_default_remote_args(ray_remote_args: Dict[str, Any]) -> Dict[str, Any
...
+        if (
+            "max_retries" not in ray_remote_args
```
Comment (c21) [inline python/ray/data/_internal/execution/operators/actor_pool_map_operator.py:300]: why it's not `max_task_retries`?

File: python/ray/data/_internal/execution/operators/actor_pool_map_operator.py (hunk 2 above)
Comment (ericl) [reply python/ray/data/_internal/execution/operators/actor_pool_map_operator.py:298]: My thoughts here were that for a long running job, there could be an indefinite number of node failures. Hence, the number of actor restarts should be unbounded. If there is a bug, then the task retry limit will eventually raise the error.

**Review by c21 — APPROVED** [review]

File: python/ray/data/_internal/execution/operators/actor_pool_map_operator.py (hunk 2 above)
Comment (jianoaix) [reply python/ray/data/_internal/execution/operators/actor_pool_map_operator.py:298]: Shall we update the documentation?
Maybe we can do separately, as I checked the docs it looks we don't have an explanation about Dataset FT yet.

**Review by jianoaix — APPROVED** [review]

**Pull request merged**